File: Adapter/Adapter.cpp
#include <iostream>

// Legacy class with incompatible interface
class LegacyPrinter {
public:
    void printDocument() {
        std::cout << "Printing document..." << std::endl;
    }
};

// Modern interface
class Printer {
public:
    virtual void print() = 0;
    virtual void scan() = 0;
};

// Adapter to make LegacyPrinter compatible with Printer interface
class LegacyPrinterAdapter : public Printer {
private:
    LegacyPrinter* legacyPrinter;

public:
    LegacyPrinterAdapter(LegacyPrinter* legacyPrinter) : legacyPrinter(legacyPrinter) {}

    void print() override {
        legacyPrinter->printDocument();
    }

    void scan() override {
        std::cout << "Scanning document..." << std::endl;
    }
};

int main() {
    LegacyPrinter legacyPrinter;
    LegacyPrinterAdapter adapter(&legacyPrinter);

    // Using the adapted LegacyPrinter as a Printer
    Printer* printer = &adapter;

    printer->print(); // Uses the adapted print method
    printer->scan();  // Uses the new scan method

    return 0;
}

File: Adapter/Adapter.java
// Legacy class with incompatible interface
class LegacyPrinter {
    void printDocument() {
        System.out.println("Printing document...");
    }
}

// Modern interface
interface Printer {
    void print();
    void scan();
}

// Adapter to make LegacyPrinter compatible with Printer interface
class LegacyPrinterAdapter implements Printer {
    private LegacyPrinter legacyPrinter;

    public LegacyPrinterAdapter(LegacyPrinter legacyPrinter) {
        this.legacyPrinter = legacyPrinter;
    }

    @Override
    public void print() {
        legacyPrinter.printDocument();
    }

    @Override
    public void scan() {
        System.out.println("Scanning document...");
    }
}

// Client code
public class Main {
    public static void main(String[] args) {
        LegacyPrinter legacyPrinter = new LegacyPrinter();
        Printer printer = new LegacyPrinterAdapter(legacyPrinter);
        
        printer.print(); // Uses the adapted print method
        printer.scan();  // Uses the new scan method
    }
}
